lib/src/rules/invariant_boolean.dart

```diff
@@ -0,0 +1,254 @@
+if(foo == bar && something && foo != bar) {...}
+```
+
+**BAD:**
+```
+void compute(int foo) {
+  if (foo == 4) {
+    doSomething();
+    // We know foo is equal to 4 at this point, so the next condition is always false
+    if (foo > 4) {...}
+    ...
+  }
+  ...
+}
+```
+
+**BAD:**
+```
+void compute(boolean foo) {
+  if (foo) {
+    return;
+  }
+  doSomething();
+  // foo is always false here
+  if (foo){...}
+  ...
+}
+```
+
+**GOOD:**
+```
+void nestedOK() {
+  if (foo == bar) {
+    foo = baz;
+    if (foo != bar) {...}
+  }
+}
+```
+
+**GOOD:**
+```
+void nestedOk2() {
+  if (foo == bar) {
+    return;
+  }
+
+  foo = baz;
+  if (foo == bar) {...} // OK
+}
+```
+
+**GOOD:**
+```
+void nestedOk5() {
+  if (foo != null) {
+    if (bar != null) {
+      return;
+    }
+  }
+
+  if (bar != null) {...} // OK
+}
+```
+''';
+
+
+class InvariantBoolean extends LintRule {
+  _Visitor _visitor;
+
+  InvariantBoolean() : super(
+      name: 'invariant_boolean',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  visitBinaryExpression(BinaryExpression node) {
+    // Right part discards reporting a subexpression already reported.
+    if (node.bestType.name != 'bool' || node.parent is! IfStatement) {
+      return;
+    }
+
+    TestedExpressions testedNodes = _findPreviousTestedExpressions(node);
+    testedNodes.evaluateInvariant().forEach((ContradictoryComparisons e) {
+      _ContradictionReportRule reportRule = new _ContradictionReportRule(e);
+      reportRule.reporter = rule.reporter;
+      reportRule.reportLint(e.second);
+    });
+
+    // In dart booleanVariable == true is a valid comparison since the variable
+    // can be null.
+    if (!BooleanExpressionUtilities.EQUALITY_OPERATIONS
+        .contains(node.operator.type) &&
+        (node.leftOperand is BooleanLiteral ||
+            node.rightOperand is BooleanLiteral)) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    if (node.condition is BooleanLiteral) {
+      rule.reportLint(node.condition);
+    }
+  }
+}
+
+/// The only purpose of this rule is to report the second node on a cotradictory
+/// comparison indicating the first node as the cause of the inconsistency.
+class _ContradictionReportRule extends LintRule {
+  _ContradictionReportRule(ContradictoryComparisons comparisons) : super(
+      name: 'invariant_boolean',
+      description: _desc + ' verify: ${comparisons.first}.',
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental);
+}
+
+TestedExpressions _findPreviousTestedExpressions(BinaryExpression node) {
+  Block block = node.parent
```

Comment:
It doesn't really matter, but I don't think there's any reason to get the `parent` at this point because a binary expression cannot be a direct child of a block.